google/cloud/bigtable/internal/data_connection_impl.cc

```diff
@@ -619,8 +619,29 @@ DataConnectionImpl::AsyncReadRow(std::string const& table_name,
 }
 
 StatusOr<bigtable::PreparedQuery> DataConnectionImpl::PrepareQuery(
-    bigtable::PrepareQueryParams const&) {
-  return Status(StatusCode::kUnimplemented, "not implemented");
+    bigtable::PrepareQueryParams const& params) {
+  auto current = google::cloud::internal::SaveCurrentOptions();
+  google::bigtable::v2::PrepareQueryRequest request;
+  request.set_instance_name(params.instance.FullName());
+  request.set_app_profile_id(app_profile_id(*current));
+  request.set_query(params.sql_statement.sql());
+  for (auto const& p : params.sql_statement.params()) {
+    (*request.mutable_param_types())[p.first] = p.second.type();
+  }
+  auto response = google::cloud::internal::RetryLoop(
+      retry_policy(*current), backoff_policy(*current),
+      Idempotency::kNonIdempotent,
```

Comment:
We should check with the Bigtable team to see if we're correct in treating this RPC as non-idempotent.

Reply:
Hm that's a good point, will confirm if this the desired behavior (created https://github.com/googleapis/google-cloud-cpp/issues/15663 for an immediate follow up)